examples/schroedinger-splitting/schroedinger-splitting.cpp

```diff
@@ -0,0 +1,204 @@
+std::pair<cv::VideoWriter, cv::Mat> build_output(int n, double fps)
+{
+    cv::Size videosize{n, n};
+    auto output =
+        std::make_pair(cv::VideoWriter{}, cv::Mat{videosize, CV_8UC3});
+    auto fourcc = cv::VideoWriter::fourcc('a', 'v', 'c', '1');
+    output.first.open("nls.mp4", fourcc, fps, videosize);
+    return output;
+}
+
+
+// Write the current frame to video output using the above color mapping
+void output_timestep(std::pair<cv::VideoWriter, cv::Mat> &output, int n,
+                     const std::complex<double> *data)
+{
+    for (int i = 0; i < n; i++) {
+        auto row = output.second.ptr(i);
+        for (int j = 0; j < n; j++) {
+            set_val(&row[3 * j], abs(data[i * n + j]));
+        }
+    }
+    output.first.write(output.second);
+}
+
+
+int main(int argc, char *argv[])
+{
+    using vec = gko::matrix::Dense<std::complex<double>>;
+    using real_vec = gko::matrix::Dense<double>;
+    using fft2 = gko::matrix::Fft2;
+
+    // Problem parameters:
+    // simulation length
+    const auto t0 = 15.0;
+    // scaling factor for non-linearity
+    const auto nonlinear_scale = 1.0;
+    // scaling factor for potential
+    const auto potential_scale = 3.0;
+    // Simulation parameters:
+    // time scaling factor
+    const auto time_scale = 0.25;
+    // number of grid points in each dimension
+    const auto n = 256;
+    // number of simulation steps per second
+    const auto steps_per_sec = 1000;
+    // number of video frames per second
+    const auto fps = 25;
+    // number of grid points
+    const auto n2 = n * n;
+    // phase difference between neighboring grid points
+    const auto h = 2.0 * gko::pi<double>() / n;
+    const auto h2 = h * h;
+    // time step size for the simulation
+    const auto tau = 1.0 / steps_per_sec;
+    const auto idx = [&](int i, int j) { return i * n + j; };
+
+    // create an OpenMP executor
+    auto exec = gko::OmpExecutor::create();
+    // load initial state vector
+    std::ifstream initial_stream("data/gko_logo_2d.mtx");
+    std::ifstream potential_stream("data/gko_text_2d.mtx");
+    auto amplitude = gko::read<vec>(initial_stream, exec);
+    auto potential = gko::read<real_vec>(potential_stream, exec);
+    // create vector for frequency space representation
+    auto frequency = vec::create(exec, amplitude->get_size());
+    // create Fourier matrix
+    auto fft = fft2::create(exec, n, n);
+    auto ifft = fft->conj_transpose();
+    // prepare video output
+    auto output = build_output(n, fps);
+    // time stamp of the last output frame (sentinel value)
+    double last_t = -t0;
+    // execute splitting method: time step in linear part, then non-linear part
+    for (double t = 0; t < t0; t += tau) {
+        // if enough time has passed, output the next frame
+        if (t - last_t > 1.0 / fps) {
+            last_t = t;
+            std::cout << t << std::endl;
+            output_timestep(output, n, amplitude->get_const_values());
```

Comment:
Due to the point-wise operations for the non-linear part of the equation, we cannot do this without implementing a custom kernel. I would like to avoid this for now, but my simplified kernel launch example from a while ago would make this pretty simple.

Reply:
yes, exactly. We'd need to move the common kernel interface into the public interface and add some CMake magic to compile a separate file for all available backends, but then this would be possible as well.